runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -174,24 +183,29 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
     return runner.apply(transform, input);
   }
 
-  Optional<Boolean> checkForSuccess(DataflowPipelineJob job)
+  /**
+   * Check success status of a dataflow job with given job metrics.
+   *
+   * @return Optional.of(false) if job failed/cancelled or PAssert failed,
+   *         Optional.of(true) if number of success PAssert meet expects,
+   *         or Optional.absent() otherwise.
+   */
+  Optional<Boolean> checkForSuccess(DataflowPipelineJob job, JobMetrics metrics)
       throws IOException {
     State state = job.getState();
     if (state == State.FAILED || state == State.CANCELLED) {
       LOG.info("The pipeline failed");
       return Optional.of(false);
     }
 
-    JobMetrics metrics = options.getDataflowClient().projects().jobs()
-        .getMetrics(job.getProjectId(), job.getJobId()).execute();
-
     if (metrics == null || metrics.getMetrics() == null) {
       LOG.warn("Metrics not present for Dataflow job {}.", job.getJobId());
```

Comment:
This is not about "return value" v.s else block.

My suggestion is to put the preconditions to the top of the function. So, the rest of the code doesn't need to have multiple level of code blocks.

http://c2.com/cgi/wiki?GuardClause

For the actual logic (returning), I personally prefer to have full "if () {} else {} blocks". But, that might be controversial, and case by case.